[13.0] Add module sale_cutoff_time_weekday_delivery

grindtildeath opened the pull request:
Depends on:
 - [ ] #1101 
 - [ ] #1104 

This module adds some unittests to ensure Method Resolution Order of
`sale.order.line._prepare_procurement_values` for modules
`sale_cutoff_time_delivery` and `sale_partner_delivery_window` is correct.

grindtildeath committed:
Add module sale_cutoff_time_weekday_delivery

grindtildeath committed:
fixup! Add module sale_cutoff_time_weekday_delivery

grindtildeath committed:
fixup! fixup! Add module sale_cutoff_time_weekday_delivery

grindtildeath committed:
fixup! fixup! fixup! Add module sale_cutoff_time_weekday_delivery

jbaudoux reviewed (changes requested):
We need to ensure that cutoff computation is applied before time window. If after cutoff, we add a day. Then, if not in the time window, we add additional days.
When I tested, it executed in the contrary order

grindtildeath committed:
fixup! fixup! fixup! fixup! Add module sale_cutoff_time_weekday_delivery

grindtildeath committed:
[IMP] sale_partner_cutoff_delivery_window: black, isort

grindtildeath committed:
Readme

grindtildeath commented:
@jbaudoux This test (https://github.com/OCA/sale-workflow/pull/1107/files#diff-4d050161f1bf44a1ee247ea0e214f592R194-R230) must ensure MRO is correct. Can you check if you still have issues after the refresh of submodules?

grindtildeath commented on sale_partner_cutoff_delivery_window/__manifest__.py:
```diff
@@ -0,0 +1,16 @@
+    "application": False,
+    "installable": True,
+    "depends": ["sale_cutoff_time_delivery", "sale_partner_delivery_window"],
+    "auto-install": True,
```
```suggestion
    "auto_install": True,
```

grindtildeath committed:
Update sale_partner_cutoff_delivery_window/__manifest__.py

jgrandguillaume reviewed (approved):
Functional test, work great

Commit:
Apply cut-off time after delivery time window

When a customer has a delivery time window, we want to apply the cut-off
time on the scheduded date.

Whenever we have a customer with weekly delivery, the scheduled datetime
of the delivery must be set at the cut-off time for this order:

* e.g. I set a customer to be shipped on Tuesday and Thursday between 13:55-14:30,
* I place a SO on the 29th of November 2020, my scheduled datetime on my
  delivery must be 1st december at 09:00
* 1st december because it is the next Tuesday from the order creation
* 09:00 because my cut-off time for WH is set at 09:00 and the customer
  doesn't have any specific cut-off time

simahawk commented:
replaced by #1379
